academic_observatory_workflows/workflows/crossref_events_telescope.py

```diff
@@ -12,383 +12,859 @@
-        period = pendulum.period(start_date, end_date)
-        for dt in period.range("days"):
-            date_str = dt.strftime("%Y-%m-%d")
-            start_date = date_str
-            end_date = date_str
-
-            events_url = (
-                f"https://api.eventdata.crossref.org/v1/events?mailto={self.mailto}"
-                f"&from-collected-date={start_date}&until-collected-date={end_date}&rows=1000"
-            )
-            edited_url = (
-                f"https://api.eventdata.crossref.org/v1/events/edited?mailto={self.mailto}"
-                f"&from-updated-date={start_date}&until-updated-date={end_date}&rows=1000"
-            )
-            deleted_url = (
-                f"https://api.eventdata.crossref.org/v1/events/deleted?mailto={self.mailto}"
-                f"&from-updated-date={start_date}&until-updated-date={end_date}&rows=1000"
+        self.mailto = mailto
+        self.primary_key = primary_key
+        self.observatory_api_conn_id = observatory_api_conn_id
+
+        # Wait for the previous DAG run to finish to make sure that
+        # changefiles are processed in the correct order
+        external_task_id = "dag_run_complete"
+        self.add_operator(
+            PreviousDagRunSensor(
+                dag_id=self.dag_id,
+                external_task_id=external_task_id,
+                execution_delta=timedelta(days=7),  # To match the @weekly schedule_interval
             )
+        )
 
-            urls.append(events_url)
-            if not self.first_release:
-                urls.append(edited_url)
-                urls.append(deleted_url)
-        return urls
+        self.add_setup_task(self.check_dependencies)
+        self.add_setup_task(self.fetch_releases)
+        self.add_task(self.create_datasets)
+
+        # Create snapshots of main table in case we mess up
+        # This is done before updating the tables to make sure that the snapshots haven't expired before the tables
+        # are updated
+        self.add_task(self.bq_create_main_table_snapshot)
+
+        # Download and transform data
+        self.add_task(self.download)
+        self.add_task(self.upload_downloaded)
+        self.add_task(self.transform_snapshot)
+        self.add_task(self.transform_created_edited)
+        self.add_task(self.transform_deleted)
+        self.add_task(self.upload_transformed)
+
+        # Upsert records
+        self.add_task(self.bq_load_main_table)
+        self.add_task(self.bq_load_upsert_table)
+        self.add_task(self.bq_upsert_records)
+
+        # Delete records
+        self.add_task(self.bq_load_delete_table)
+        self.add_task(self.bq_delete_records)
+
+        # Add release info to API and cleanup
+        self.add_task(self.add_new_dataset_releases)
+        self.add_task(self.cleanup)
+
+        # The last task that the next DAG run's ExternalTaskSensor waits for.
+        self.add_operator(
+            DummyOperator(
```

Comment:
Why use a dummy here over the cleanup task?

Reply:
I guess it isn't strictly necessary, but it makes it easier to test the workflows because the cleanup task can be commented out during testing.

Reply:
Ah I see. Fair enough